### PR TITLE
doc: add operational flow figure for netdev

### PR DESCRIPTION
**Pull request by miri64**
I made this graphic for my thesis and I (and @OlegHahm) think it might be nice to have this in the doc.

### CODE REVIEW

File: drivers/include/net/netdev2.h [hunk 1]
```diff
@@ -20,13 +20,19 @@
...
  *
- * 1. packet arrives -> driver calls netdev2->event_callback with
- *    event==NETDEV_EVENT_ISR (from interrupt routine)
- * 2. event_callback wakes up upper layer thread
- * 3. upper layer calls netdev2->driver.isr()
- * 4. netdev2->driver.isr() calls netdev2->event_callback() with
- *    event==NETDEV_EVENT_RX_COMPLETE
- * 5. netdev2->event_callback() uses netdev2->driver.recv() to fetch packet
+ * 1. packet arrives for device
+ * 2. driver calls @ref netdev2_t::event_callback "netdev->event_callback()"
```
Comment (A-Paul) [inline drivers/include/net/netdev2.h]: I your sequential diagram the *ISR* is calling `netdev::event_cb()`

File: drivers/include/net/netdev2.h [hunk 2]
```diff
@@ -20,13 +20,19 @@
...
+ *    with `event` := @ref NETDEV2_EVENT_ISR (from Interrupt Service Routine)
+ *    and wakes up event handler
+ * 3. event handler calls @ref netdev2_driver_t::isr "netdev2->driver->isr()"
```
Comment (A-Paul) [inline drivers/include/net/netdev2.h]: Named `netdev::driver::isr()` in the diagram.

File: drivers/include/net/netdev2.h [hunk 3]
```diff
@@ -20,13 +20,19 @@
...
+ *    (from thread context)
+ * 4. @ref netdev2_driver_t::isr "netdev->driver->isr()" calls
+ *    @ref netdev2_t::event_callback "netdev->event_callback()" with
```
Comment (A-Paul) [inline drivers/include/net/netdev2.h]: Also different naming.

File: drivers/include/net/netdev2.h [hunk 4]
```diff
@@ -20,13 +20,19 @@
...
+ *    `event` := @ref NETDEV2_EVENT_RX_COMPLETE
+ * 5. @ref netdev2_t::event_callback "netdev->event_callback()" uses
+ *    @ref netdev2_driver_t::recv "netdev2->driver->recv()" to fetch packet
```
Comment (A-Paul) [inline drivers/include/net/netdev2.h]: ...

File: drivers/include/net/netdev2.h (hunk 2 above)
Comment (miri64) [inline drivers/include/net/netdev2.h]: It's the same here. I just use the more "C style" `->` operator here instead of the C++/UML style `::` to be more in line with the code

File: drivers/include/net/netdev2.h (hunk 3 above)
Comment (miri64) [inline drivers/include/net/netdev2.h]: See comments above.

File: drivers/include/net/netdev2.h (hunk 2 above)
Comment (miri64) [inline drivers/include/net/netdev2.h]: (The `@ref struct::member "link-text"` must be used, otherwise doxygen won't display the links correctly)

**Comment by miri64** [comment]
@A-Paul ping?

File: drivers/include/net/netdev2.h (hunk 1 above)
Comment (A-Paul) [inline drivers/include/net/netdev2.h]: I'm not sure if I don't understand your answer or if my comment was misleading.
I wanted to point out that there are differences between the text and the diagram. You're writing:
> 2. driver calls netdev->event_callback() ....

But in the diagram the *ISR* is calling the function (As far as I can see).

Comment (miri64) [inline drivers/include/net/netdev2.h]: Ah, yes this is indeed missleading (it's the ISR *registered by* the driver, but never the less you are right).

Comment (miri64) [inline drivers/include/net/netdev2.h]: Mh but reading the list item again in total it is mostly clear IMHO

> driver calls netdev->event_callback() with `event` := NETDEV2_EVENT_ISR ***(from Interrupt Service Routine)*** and wakes up event handler

Any idea how to make it more clear?

Comment (A-Paul) [inline drivers/include/net/netdev2.h]: In general I would refer to what I remember about basic principles of documentation.

  - *Try to avoid information that is not essential for understanding the specific topic.*
  - *Use a figure only to provide additional structural overview. Don't repeat information which is already in your text. Instead try to use references to the explaining passages.*  
  Have you considered just using the sequence numbers as message labels in the SD? That is space saving too.
  - *Always remember that the reader is someone who needs your guide through the topic. And not someone who is capable to resolve inconsistencies on its own.*  
   You claimed the different notations to be all correct. Indeed it is true.  
   But if **I** see different notaions, slightly different function names, etc. Then I try to figure out if the is some meaning in those different representations or I check if I'm the wrong part of the diagram. For that moment I'm completely distracted from the sequence of events you want to show me.

Phew, that turned into something like the "Wort zum Sonntag" :D
Not intended to offend you. I just want to make clear why I'm stressing those *stupid questions* so much.

>> driver calls netdev->event_callback() with `event` := NETDEV2_EVENT_ISR ***(from Interrupt Service Routine)*** and wakes up event handler

> Any idea how to make it more clear?

Just try what you answered to my questions to get me on the track.

Before the start of the sequence, mention that the ISR has to be registered by the driver as a precondition.
And then e.g.:
1. ISR calls netdev->event_callback(NETDEV2_EVENT_ISR) and wakes up event handler

Given that I understood what is going on...

**Comment by A-Paul** [comment]
@miri64, slighly OT. Which tool  do you use to create sequence diagrams? Are you aware of [sdedit](http://sdedit.sourceforge.net/)?

File: drivers/include/net/netdev2.h (hunk 1 above)
Comment (miri64) [inline drivers/include/net/netdev2.h]: Never understood as offence, just frustrated with my own incompetence ;-). Will try to take your advice to heart and fix this... tomorrow ;-)

**Comment by miri64** [comment]
> @miri64, slighly OT. Which tool do you use to create sequence diagrams? Are you aware of sdedit?

Don't laugh: hand-crafted SVG in vim ;-)

**Comment by miri64** [comment]
(how else did you think the XML came out that neat :P)

**Comment by miri64** [comment]
Addressed comments.

**Comment by A-Paul** [comment]
> Don't laugh: hand-crafted SVG in vim ;-)

Regarding the font definition block this is outstanding.

**Comment by miri64** [comment]
> Regarding the font definition block this is outstanding. :D

Well I used inkscape as a starting point for formatting (and if I remember correctly a UML editor to get a structural template), but I found that generated XML is much harder to version control than hand-edited, so I thought why not make it proper so that most changes are easier to review.

**Comment by A-Paul** [comment]
> Addressed comments.

Very close to what I imagined. :) Thanks.
ACK.

**Comment by miri64** [comment]
Squashed.

**Comment by A-Paul** [comment]
> Squashed.

Sure? :)

**Comment by miri64** [comment]
Yes. One commit for the introduction of the graphic, one for the netdev doc change.

**Comment by A-Paul** [comment]
Somehow I thought that, in the end, there should be only one commit in a PR.
